### PR TITLE
Addding kfto component to odh operator

### PR DESCRIPTION
**Pull request by tedhtchang**
ref: https://issues.redhat.com/browse/RHOAIENG-4923
ref: https://issues.redhat.com/browse/RHOAIENG-4663

## Description
Adding kubeflow training-operator component to the odh-operator

## How Has This Been Tested?
Manually tested with @z103cb 's [rhoai-manifests branch](https://github.com/z103cb/training-operator/tree/rhoai-manifests/manifests) on [CRC](https://developers.redhat.com/products/openshift-local/overview)
```
# clone this PR
git clone --single-branch --branch add-kfto-component https://github.com/tedhtchang/opendatahub-operator.git
cd opendatahub-operator

# Download manifests, this will copy kfto manifests from the [rhoai-manifests branch](https://github.com/z103cb/training-operator/tree/rhoai-manifests/manifests)
make get-manifests

# Build image from local
export IMG="<your img reg>/<your org>/opendatahub-operator:dev"
make -e IMAGE_BUILD_FLAGS="--build-arg USE_LOCAL=true" -e IMG=${IMG} image-build
docker push ${IMG}

# deploy
export OPERATOR_NAMESPACE=opendatahub 
make deploy -e IMG=${IMG} -e OPERATOR_NAMESPACE=opendatahub-operator-system



cat <<EOF | oc apply -f -
kind: DSCInitialization
apiVersion: dscinitialization.opendatahub.io/v1
metadata:
  name: default-dsci
spec:
  applicationsNamespace: opendatahub
  monitoring:
    managementState: Managed
    namespace: opendatahub
  serviceMesh:
    controlPlane:
      metricsCollection: Istio
      name: data-science-smcp
      namespace: istio-system
    managementState: Managed
  trustedCABundle:
    customCABundle: ''
    managementState: Managed
---
apiVersion: datasciencecluster.opendatahub.io/v1
kind: DataScienceCluster
metadata:
  name: default-dsc
spec:
  components:
    codeflare:
      managementState: Removed
    dashboard:
      managementState: Removed
    datasciencepipelines:
      managementState: Removed
    kserve:
      managementState: Removed
      serving:
        ingressGateway:
          certificate:
            type: SelfSigned
        managementState: Managed
        name: knative-serving
    kueue:
      managementState: Removed
    modelmeshserving:
      managementState: Removed
    modelregistry:
      managementState: Removed
    ray:
      managementState: Removed
    trustyai:
      managementState: Removed
    workbenches:
      managementState: Removed
    trainingoperator:
      managementState: Managed
EOF
```
Observe components pod been created
image
Clean up
```
oc delete dsc default-dsc
oc delete dsci default-dsci
make undeploy
```

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/trainingoperator/trainingoperator.go
Comment (zdtsw) [inline components/trainingoperator/trainingoperator.go:59]: double check here, 
but i think it is `RELATED_IMAGE_ODH_TRAINING_OPERATOR_IMAGE`

File: get_all_manifests.sh
Comment (zdtsw) [inline get_all_manifests.sh:22]: we should wait manifest PR gets into `opendathub-io` first  to dev branch and remove `z103cb`+` rhoai-manifests`

**Review by zdtsw — COMMENTED** [review]
we will need some more to cover test e2e part 
otherwise i think the draft look good + some comments left to check

File: get_all_manifests.sh
Comment (tedhtchang) [reply get_all_manifests.sh:22]: yes. let's make this line the last thing to update.

**Comment by jbusche** [comment]
I got it to deloy @tedhtchang following your instructions on an OC 4.14.17 cluster.  I had to adjust one thing, during the creation of the dsci and dsc there's an extra `cat <<EOF | oc apply -f -` in the middle that I had to remove in order for them to install fully.

**Comment by jbusche** [comment]
I'm trying the `Deploying operator using OLM` instructions from [here](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/README.md#deployment) and I ran into a few problems.

Problem 1 - it's trying to pull busybox from docker.io, which is a problem on my cluster.  If I quick oc edit the pod, I can fix it before the run bundle errors out
Problem 2 - my opendatahub-operator-controller-manager-6877656bbd-xdkhg pod is in a Status of CrashLoopBackOff, seemingly because it's missing kfdef:
```
oc logs -f opendatahub-operator-controller-manager-6877656bbd-xdkhg
....
2024-03-30T06:12:52.806Z	ERROR	controller-runtime.source	if kind is a CRD, it should be installed before calling Start	{"kind": "KfDef.kfdef.apps.kubeflow.org", "error": "no matches for kind \"KfDef\" in version \"kfdef.apps.kubeflow.org/v1\""}
2024-03-30T06:12:52.807Z	ERROR	controller.kfdef-controller	Could not wait for Cache to sync	{"reconciler group": "kfdef.apps.kubeflow.org", "reconciler kind": "KfDef", "error": "failed to wait for kfdef-controller caches to sync: no matches for kind \"KfDef\" in version \"kfdef.apps.kubeflow.org/v1\""}
```

Here's the steps I followed:
```
Trying the OLM bundle

Step 1.  Clone the PR
git clone --single-branch --branch add-kfto-component https://github.com/tedhtchang/opendatahub-operator.git
cd opendatahub-operator

Step 2. I pull down all the manifests
make get-manifests

Step 3. Login to quay
podman login quay.io -u jbusche

Step 4. make the bundle
export OPERATOR_NAMESPACE=opendatahub
make bundle

Step 5. push the bundle
make bundle-build bundle-push BUNDLE_IMG=quay.io/jbusche/opendatahub-operator-bundle:tedv1

Step 6. Make sure it's readable/public on quay.io
https://quay.io/repository/jbusche/opendatahub-operator-bundle

Step 7. Run the Bundle on a cluster:
operator-sdk run bundle quay.io/jbusche/opendatahub-operator-bundle:tedv1 --namespace $OPERATOR_NAMESPACE

Step 8. Quick in another window:
oc edit pod
and change docker.io busybox image to:
quay.io/project-codeflare/busybox:1.36

Step 9.  Look at the pods:
oc get pods
NAME                                                              READY   STATUS             RESTARTS      AGE
31a8154222982c03d9c19a1518fc58d876604a1947d58df1f0523497aafxvff   0/1     Completed          0             15m
opendatahub-operator-controller-manager-6877656bbd-xdkhg          0/1     CrashLoopBackOff   7 (86s ago)   15m
quay-io-jbusche-opendatahub-operator-bundle-tedv1                 1/1     Running            0             15m
```

**Comment by zdtsw** [comment]
> quay.io/jbusche/opendatahub-operator-bundle:tedv1

could you check if your `opendatahub-operator-controller-manager-6877656bbd-xdkhg` pod is running with this image `quay.io/jbusche/opendatahub-operator-bundle:tedv1` ?

**Comment by jbusche** [comment]
You bring up a good point @zdtsw, the csv was creating it as `quay.io/opendatahub/opendatahub-operator:latest`.  I'll try to build it locally and use the local image.

**Comment by tedhtchang** [comment]
@zdtsw @anishasthana Do you know how if this error is related to this PR and how to resolve it ?
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_opendatahub-operator/944/pull-ci-opendatahub-io-opendatahub-operator-incubation-opendatahub-operator-e2e/1774875743944183808
```
    --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap (343.55s)
        --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test (340.96s)
            --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/create_data_science_cluster (10.14s)
            --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created (60.08s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_TrustyAI (10.05s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_ModelRegistry (10.08s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Ray (10.11s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Dashboard (10.11s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Workbenches (10.11s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_CodeFlare (10.11s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_DataSciencePipelines (10.12s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_TrainingOperator (10.14s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Kserve (10.14s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Kueue (50.04s)
                --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_ModelMeshServing (120.05s)
            --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/trigger_deletion_using_labeled_config_map (0.16s)
            --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/owned_namespaces_should_be_deleted (40.05s)
            --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/dsci_should_be_deleted (0.04s)
            --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/applications_resources_should_be_deleted (110.44s)
FAIL
FAIL	github.com/opendatahub-io/opendatahub-operator/v2/tests/e2e	784.132s
FAIL
make: *** [e2e-test] Error 1
{"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"k8s.io/test-infra/prow/entrypoint/run.go:84","func":"k8s.io/test-infra/prow/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2024-04-01T20:01:36Z"}
error: failed to execute wrapped command: exit status 2 
```

File: components/component.go
Comment (tedhtchang) [inline components/component.go:136]: What should I do about this function. Looks like I could just replicate?

Comment (anishasthana) [reply components/component.go:136]: Yeah I think having a similar set of rules to what we have for kuberay is good.

**Comment by zdtsw** [comment]
> Edited: Re-ran and passed. @zdtsw @anishasthana Do you know how if this error is related to this PR and how to resolve it ? https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_opendatahub-operator/944/pull-ci-opendatahub-io-opendatahub-operator-incubation-opendatahub-operator-e2e/1774875743944183808
> 
> ```
>     --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap (343.55s)
>         --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test (340.96s)
>             --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/create_data_science_cluster (10.14s)
>             --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created (60.08s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_TrustyAI (10.05s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_ModelRegistry (10.08s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Ray (10.11s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Dashboard (10.11s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Workbenches (10.11s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_CodeFlare (10.11s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_DataSciencePipelines (10.12s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_TrainingOperator (10.14s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Kserve (10.14s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_Kueue (50.04s)
>                 --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/ensure_all_components_created/Validate_ModelMeshServing (120.05s)
>             --- FAIL: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/trigger_deletion_using_labeled_config_map (0.16s)
>             --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/owned_namespaces_should_be_deleted (40.05s)
>             --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/dsci_should_be_deleted (0.04s)
>             --- PASS: TestOdhOperator/remove_components_by_using_labeled_configmap/e2e-test/applications_resources_should_be_deleted (110.44s)
> FAIL
> FAIL	github.com/opendatahub-io/opendatahub-operator/v2/tests/e2e	784.132s
> FAIL
> make: *** [e2e-test] Error 1
> {"component":"entrypoint","error":"wrapped process failed: exit status 2","file":"k8s.io/test-infra/prow/entrypoint/run.go:84","func":"k8s.io/test-infra/prow/entrypoint.Options.internalRun","level":"error","msg":"Error executing test process","severity":"error","time":"2024-04-01T20:01:36Z"}
> error: failed to execute wrapped command: exit status 2 
> ```

yeah, the original error usually not related to your change, sometime can be glitch due to the test env.

File: components/component.go
Comment (zdtsw) [reply components/component.go:136]: for the time being your change is enough
this wont get triggered in ODH, we will need to work on RHOAI later to define such rules later

Comment (zdtsw) [reply components/component.go:136]: normally, the process is:
first add logic in ODH for component but skip the monitoring part (since it is not get called)
when it is sync down to RHOAI, we add extra monitoring part: logic + rules etc
after the code is merged into RHOAI, we do backport for the logic to ODH

you have defined the "logic" in the PR, we can keep it here. or if you perfer removing it, 
i am fine with either.

Comment (tedhtchang) [reply components/component.go:136]: @zdtsw Thanks for reviewing. I prefer keep it.

**Review by z103cb — COMMENTED** [review]
This PR looks ok to me, but I will defer to the experts.

File: get_all_manifests.sh
Comment (zdtsw) [reply get_all_manifests.sh]: i've updated this in https://github.com/opendatahub-io/opendatahub-operator/pull/944/commits/baebfc6c6255151c8655fad82c9cd4074bb8bfdc  after https://github.com/opendatahub-io/training-operator/pull/3 is merged

**Comment by zdtsw** [comment]
i did some test on the PR
i saw something strange but more likely it is from other components change made recently.
will spend a bit more time to troubleshoot but for this PR, it is fine from me

**Review by zdtsw — APPROVED** [review]

**Pull request merged**
